### PR TITLE
chore: Update DataFusion pin, upgrade `date_bin` and `InfluxQL` to use `Interval(MonthDayNano)` 

### LINKED ISSUE #7204: InfluxQL-IOx: `GROUP BY TIME` precision limited to milliseconds
InfluxQL relies on the `DATE_BIN` function when producing an aggregate query that specifies a combination of `GROUP BY TIME` and `FILL(none)`. The `stride` parameter of the `DATE_BIN` function is limited to a maximum precision of milliseconds.

The `date_bin_gapfill` function will need to be amended to support `MonthDayNano` for the stride argument:

https://github.com/influxdata/influxdb_iox/blob/30fea67701d0f6c93d847a8bd3ab4dd2796af734/query_functions/src/gapfill.rs#L24

See:

https://github.com/apache/arrow-datafusion/blob/146a949218ec970784974137277cde3b4e547d0a/datafusion/expr/src/function.rs#L453

Depends on:

- https://github.com/apache/arrow-datafusion/issues/5697
- https://github.com/apache/arrow-datafusion/pull/5698

### PR DESCRIPTION
**Pull request by alamb**
Part of https://github.com/influxdata/idpe/issues/17440

# Rationale
* keep up with dependencies, and prepare for arrow 37 upgrade

# Changes:
1. Update DataFusion pin to https://github.com/apache/arrow-datafusion/commit/34c9bce0146e0ebacfb10334ee7aef13fc8cc94f
2. Update for API changes

Note this PR also fixes  #7204 because the `date_bin` function now uses `Interval(MonthDayNano)` which can precisely represent nanoseconds precision intervals. I will leave comments inline.

### CODE REVIEW

File: datafusion_util/src/sort_exprs.rs [hunk 1]
```diff
@@ -1,52 +0,0 @@
-use datafusion::{
```
Comment (alamb) [inline datafusion_util/src/sort_exprs.rs:1]: https://github.com/apache/arrow-datafusion/pull/5863 was merged so this code is no longer needed

File: influxdb_iox/tests/query_tests2/cases/in/gapfill.sql.expected [hunk 2]
```diff
@@ -4,18 +4,18 @@
 ----------
 | plan_type    | plan    |
 ----------
-| logical_plan    | Projection: date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z")) AS minute, COUNT(cpu.user)    |
-|    |   GapFill: groupBy=[[date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))]], aggr=[[COUNT(cpu.user)]], time_column=date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z")), stride=IntervalDayTime("600000"), range=Included(TimestampNanosecond(957528000000000000, None))..Included(TimestampNanosecond(957531540000000000, None))    |
-|    |     Aggregate: groupBy=[[datebin(IntervalDayTime("600000"), cpu.time, TimestampNanosecond(0, None)) AS date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))]], aggr=[[COUNT(cpu.user)]]    |
+| logical_plan    | Projection: date_bin_gapfill(IntervalMonthDayNano("600000000000"),cpu.time,Utf8("1970-01-01T00:00:00Z")) AS minute, COUNT(cpu.user)    |
```
Comment (alamb) [inline influxdb_iox/tests/query_tests2/cases/in/gapfill.sql.expected:7]: these changes reflect that the type of the `date_bin_gapfill` argument is now a higher precision interval interval type -- namely `MonthDayNano`  https://docs.rs/arrow/latest/arrow/datatypes/enum.IntervalUnit.html#variant.MonthDayNano

File: influxdb_iox/tests/query_tests2/sql_errors.rs [hunk 3]
```diff
@@ -10,7 +10,7 @@ async fn schema_merge_nonexistent_column() {
         setup_name: "MultiChunkSchemaMerge",
         sql: "SELECT * from cpu where foo = 8",
         expected_error_code: tonic::Code::InvalidArgument,
-        expected_message: r#"Error while planning query: Schema error: No field named "foo". Valid fields are "cpu"."host", "cpu"."region", "cpu"."system", "cpu"."time", "cpu"."user"."#,
+        expected_message: r#"Error while planning query: Schema error: No field named foo. Valid fields are cpu.host, cpu.region, cpu.system, cpu.time, cpu.user."#,
```
Comment (alamb) [inline influxdb_iox/tests/query_tests2/sql_errors.rs:13]: Due to https://github.com/apache/arrow-datafusion/pull/5778

File: iox_query_influxql/src/plan/planner_time_range_expression.rs [hunk 4]
```diff
@@ -74,32 +73,9 @@ pub(in crate::plan) fn time_range_to_df_expr(expr: &Expr, tz: Option<chrono_tz::
 /// Simplifies `expr` to an InfluxQL duration and returns a DataFusion interval.
 ///
 /// Returns an error if `expr` is not a duration expression.
-///
```
Comment (alamb) [inline iox_query_influxql/src/plan/planner_time_range_expression.rs:77]: FYI @stuartcarnie  -- I upgraded the code to use an interval time that has sufficient precision to represent nanosecond precision (which was also required in order to call date_bin)

File: query_functions/src/gapfill.rs [hunk 5]
```diff
@@ -40,7 +40,7 @@ pub(crate) static DATE_BIN_GAPFILL: Lazy<Arc<ScalarUDF>> = Lazy::new(|| {
     Arc::new(create_udf(
         DATE_BIN_GAPFILL_UDF_NAME,
         vec![
-            DataType::Interval(IntervalUnit::DayTime),       // stride
+            DataType::Interval(IntervalUnit::MonthDayNano), // stride
```
Comment (alamb) [inline query_functions/src/gapfill.rs:43]: This is to match the signature in datafusion

File: iox_query_influxql/src/plan/planner_time_range_expression.rs [hunk 6]
```diff
@@ -597,20 +572,33 @@ mod test {
...
+                ScalarValue::new_interval_mdn(0, 0, 432_000_010_000_000),
+            ),
+            (
+                "-2d10ms",
+                ScalarValue::new_interval_mdn(0, 0, -172800010000000),
+            ),
+            (
+                "-2d10ns",
+                ScalarValue::new_interval_mdn(0, 0, -172800000000010),
+            ),
```
Comment (NGA-TRAN) [inline iox_query_influxql/src/plan/planner_time_range_expression.rs:592]: This ns now works 🎉

**Review by NGA-TRAN — APPROVED** [review]
LGTM

**Review by crepererum — APPROVED** [review]

**Review by wolffcm — APPROVED** [review]
Looks great, thanks for the upgrade!

**Pull request merged**